proposals/format.md

```diff
@@ -0,0 +1,320 @@
+# Efficent Params and String Formatting
+
+## Summary
+This combination of features will increase the efficiency of formatting `string` values and passing of `params` style
+arguments.
+
+## Motivation
+The allocation overhead of formatting `string` values can dominate the performance of many text based applications: 
+from the boxing penalty of `struct` types, the `object[]` allocation for `params` and the intermediate `string` 
+allocations during `string.Format` calls. In order to maintain efficiency such applications often need to abandon
+productivity features such as `params` and `string` interpolation and move to non-standard, hand coded solutions. 
+
+Consider MSBuild as an example. This is written using a lot of modern C# features by developers who are concious of 
+performance. Yet in one representative build sample MSBuild will generate 262MB of `string` allocation
+using minimal verbosity. Of that 1/2 of the allocations are short lived alloctaions inside `string.Format`. These 
+features would remove much of that on .NET Desktop and get it down to nearly zero on .NET Core due to the 
+
+The set of language features described here will enable applications to continue using these features, with very
+little or no churn to their application code base, while removing the unintended allocation overhead in the majority of 
+cases.
+
+## Detailed Design 
+There are a set of features that will be used here to achieve these results:
+
+- Expanding `params` to support a broader set of collection types.
+- Allowing for developers to customize how `string` interpolation is achieved. 
+- Allowing for interpolated `string` to bind to more efficient `string.Format` overloads.
+
+### Extending params
+The language will allow for `params` in a method signature to have the types `Span<T>`, `ReadOnlySpan<T>` and 
+`IEnumerable<T>`. The same rules for invocation will apply to these new types that apply to `params T[]`:
```

Comment:
if we're going to go to this, wouldit be oke to also allow going to ICollection/IList/IReadOnlyList at teh same time (effectively, all the types that T[] already implements)?  It feels like it would be a prime time to just roll this in since we're working in this area.

Reply:
Why would you need to support the others, since they all implement `IEnumerable<T>`? Generally speaking, types that implement things like `ICollection` or `IList` can also efficiently implement their iterators (and do in the case of most `CoreFX` types).

Reply:
I think that the currently proposed types cover most of the scenarios. That is, `Span<T>`/`ROSpan<T>` make sense since they provide allocation-less calls (and declare the immutability of the input data) and `IEnumerable<T>` allows easier interop with LINQ (it avoids the need to realize a query).

Reply:
> Why would you need to support the others, since they all implement IEnumerable<T>? 

Because i want to expose *this* api to users:

```c#
public static void Foo(params IReadOnlyList<T> whatever) { }
```

Allowing people to just pass their *existing* list to me.   Or, they can pass raw values, and it will be an array, an that doesn't need to be converted into an IReadOnlyList since it already is one.

Reply:
I'll add it to the open issues. We didn't cover it in LDM.

Reply:
Thanks so much!  I would understand if this didn't happen.  However, it's been one of those language niggles for ages.   If it's cheap, this would be nice to have :)  It would def clean up a few of my apis. 

Note: @tannergooding and i discussed this on Gitter.  It is definitely a strange thing that the language has *baked in* understanding that an array has an implicit reference conversion to IList and its ilk.  And, yet, for 'params' i *must* specify only a `T[]` type.  It seems like it would be pretty simple to just relax that to "anything an array has reference implicit conversion to".